File: apina-core/src/test/java/fi/evident/apina/model/ModelMatchers.java
package fi.evident.apina.model;

import fi.evident.apina.model.type.ApiType;
import org.hamcrest.Description;
import org.hamcrest.Matcher;
import org.hamcrest.TypeSafeMatcher;

import java.util.Collection;

import static org.hamcrest.Matchers.containsInAnyOrder;

public final class ModelMatchers {

    private ModelMatchers() {
    }

    public static Matcher<PropertyDefinition> property(String name, ApiType type) {
        return new TypeSafeMatcher<PropertyDefinition>() {
            @Override
            protected boolean matchesSafely(PropertyDefinition item) {
                return name.equals(item.getName()) && type.equals(item.getType());
            }

            @Override
            public void describeTo(Description description) {
                description.appendText("PropertyDefinition{name=")
                        .appendValue(name)
                        .appendText(", type=")
                        .appendValue(type)
                        .appendText("}");
            }
        };
    }

    @SafeVarargs
    public static Matcher<? super Collection<PropertyDefinition>> hasProperties(Matcher<PropertyDefinition>... matchers) {
        return containsInAnyOrder(matchers);
    }
}

File: apina-core/src/test/kotlin/fi/evident/apina/model/ModelMatchers.kt
package fi.evident.apina.model

import fi.evident.apina.model.type.ApiType
import org.hamcrest.Description
import org.hamcrest.Matcher
import org.hamcrest.Matchers.containsInAnyOrder
import org.hamcrest.TypeSafeMatcher

object ModelMatchers {

    fun property(name: String, type: ApiType): Matcher<PropertyDefinition> = object : TypeSafeMatcher<PropertyDefinition>() {
        override fun matchesSafely(item: PropertyDefinition) =
                name == item.name && type == item.type

        override fun describeTo(description: Description) {
            description.appendText("PropertyDefinition{name=")
                    .appendValue(name)
                    .appendText(", type=")
                    .appendValue(type)
                    .appendText("}")
        }
    }

    fun hasProperties(vararg matchers: Matcher<PropertyDefinition>): Matcher<in Collection<PropertyDefinition>> =
            containsInAnyOrder(*matchers)
}
